File: frontend/apps/ios/Pantopus/Features/Feed/FeedPreferencesSheet.swift
//
//  FeedPreferencesSheet.swift
//  Pantopus
//
//  Pulse Preferences — the bottom sheet behind the Pulse header's
//  preferences control. Mirrors RN
//  `src/components/feed/FeedPreferencesSheet.tsx`: two Place-feed toggles
//  (deals, safety alerts) plus a single political-content toggle that
//  writes both the Place and Connections columns.
//  Backed by `GET`/`PUT /api/posts/feed-preferences`
//  (`backend/routes/posts.js:2257` / `:2286`).
//

import Foundation
import Logging
import Observation
import SwiftUI

/// Render state for the Pulse preferences sheet.
public enum FeedPreferencesState: Sendable, Equatable {
    case loading
    case loaded(FeedPreferencesDTO)
    case error(message: String)
}

/// Loads + patches the signed-in user's feed preferences.
@Observable
@MainActor
public final class FeedPreferencesViewModel {
    public private(set) var state: FeedPreferencesState = .loading
    /// True while a `PUT` is in flight — disables the rows.
    public private(set) var isSaving = false
    /// Transient error banner text.
    public var toastMessage: String?

    private let api: APIClient
    private let logger = Logger(label: "app.pantopus.ios.FeedPreferences")

    init(api: APIClient = .shared) {
        self.api = api
    }

    /// `GET /api/posts/feed-preferences`.
    public func load() async {
        if case .loaded = state { return }
        await refresh()
    }

    /// Retry / re-fetch.
    public func refresh() async {
        state = .loading
        do {
            let response: FeedPreferencesResponse = try await api.request(
                FeedActionsEndpoints.feedPreferences()
            )
            state = .loaded(response.preferences)
        } catch {
            let message = (error as? APIError)?.errorDescription ?? "Couldn't load preferences."
            state = .error(message: message)
        }
    }

    /// Show / hide local business deals on the Place surface. The backend
    /// column is `hide_deals_place`, so the switch is inverted.
    public func setShowDeals(_ show: Bool) async {
        await update(FeedPreferencesUpdateRequest(hideDealsPlace: !show))
    }

    /// Show / hide safety alerts on the Place surface.
    public func setShowAlerts(_ show: Bool) async {
        await update(FeedPreferencesUpdateRequest(hideAlertsPlace: !show))
    }

    /// One switch writes both political-content columns — RN does the same
    /// (`FeedPreferencesSheet.tsx:112-115`).
    public func setShowPolitics(_ show: Bool) async {
        await update(
            FeedPreferencesUpdateRequest(
                showPoliticsConnections: show,
                showPoliticsPlace: show
            )
        )
    }

    private func update(_ body: FeedPreferencesUpdateRequest) async {
        guard case let .loaded(current) = state, !isSaving else { return }
        isSaving = true
        defer { isSaving = false }
        // Optimistic — the row reflects the tap straight away.
        state = .loaded(Self.merged(current, with: body))
        do {
            let response: FeedPreferencesResponse = try await api.request(
                FeedActionsEndpoints.updateFeedPreferences(body)
            )
            state = .loaded(response.preferences)
        } catch {
            logger.warning("Feed preference update failed: \(error)")
            state = .loaded(current)
            toastMessage = "Couldn't save that preference."
        }
    }

    private static func merged(
        _ current: FeedPreferencesDTO,
        with patch: FeedPreferencesUpdateRequest
    ) -> FeedPreferencesDTO {
        FeedPreferencesDTO(
            hideDealsPlace: patch.hideDealsPlace ?? current.hideDealsPlace,
            hideAlertsPlace: patch.hideAlertsPlace ?? current.hideAlertsPlace,
            showPoliticsConnections: patch.showPoliticsConnections ?? current.showPoliticsConnections,
            showPoliticsPlace: patch.showPoliticsPlace ?? current.showPoliticsPlace
        )
    }
}

/// Bottom sheet body for Pulse Preferences.
public struct FeedPreferencesSheet: View {
    @State private var viewModel: FeedPreferencesViewModel
    private let onClose: @MainActor () -> Void
    /// Fired after any successful write so the host feed can refetch —
    /// RN's `onPrefsChanged` → `feed.onRefresh()`.
    private let onPrefsChanged: @MainActor () -> Void

    init(
        viewModel: FeedPreferencesViewModel? = nil,
        onClose: @escaping @MainActor () -> Void,
        onPrefsChanged: @escaping @MainActor () -> Void = {}
    ) {
        _viewModel = State(initialValue: viewModel ?? FeedPreferencesViewModel())
        self.onClose = onClose
        self.onPrefsChanged = onPrefsChanged
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s0) {
            header
            content
            Spacer(minLength: Spacing.s0)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Theme.Color.appBg)
        .task { await viewModel.load() }
        .accessibilityIdentifier("pulsePreferencesSheet")
    }

    private var header: some View {
        HStack {
            Text("Pulse Preferences")
                .font(.system(size: 18, weight: .bold))
                .foregroundStyle(Theme.Color.appText)
                .accessibilityAddTraits(.isHeader)
            Spacer()
            Button(action: onClose) {
                Icon(.x, size: 18, color: Theme.Color.appTextSecondary)
                    .frame(width: 32, height: 32)
            }
            .buttonStyle(.plain)
            .accessibilityLabel("Close preferences")
            .accessibilityIdentifier("pulsePreferencesClose")
        }
        .padding(.horizontal, Spacing.s5)
        .padding(.vertical, Spacing.s3)
        .overlay(alignment: .bottom) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
        }
    }

    @ViewBuilder private var content: some View {
        switch viewModel.state {
        case .loading:
            loadingFrame
        case let .loaded(prefs):
            loadedFrame(prefs)
        case let .error(message):
            errorFrame(message: message)
        }
    }

    private var loadingFrame: some View {
        VStack(alignment: .leading, spacing: Spacing.s3) {
            ForEach(0..<3, id: \.self) { _ in
                Shimmer(height: 56, cornerRadius: Radii.sm)
            }
        }
        .padding(.horizontal, Spacing.s5)
        .padding(.top, Spacing.s4)
        .accessibilityIdentifier("pulsePreferencesLoading")
    }

    private func loadedFrame(_ prefs: FeedPreferencesDTO) -> some View {
        VStack(alignment: .leading, spacing: Spacing.s0) {
            sectionLabel("PLACE FEED")
            preferenceRow(
                title: "Show deals",
                body: "Deals and promotions from local businesses",
                isOn: !prefs.hideDealsPlace,
                identifier: "pulsePreferencesShowDeals"
            ) { next in
                await viewModel.setShowDeals(next)
                onPrefsChanged()
            }
            preferenceRow(
                title: "Show safety alerts",
                body: "Crime reports, hazards, and safety warnings",
                isOn: !prefs.hideAlertsPlace,
                identifier: "pulsePreferencesShowAlerts"
            ) { next in
                await viewModel.setShowAlerts(next)
                onPrefsChanged()
            }
            sectionLabel("CONTENT")
            preferenceRow(
                title: "Show political content",
                body: "Political posts are hidden by default to keep your feed focused"
                    + " on neighborhood life",
                isOn: prefs.showPoliticsPlace,
                identifier: "pulsePreferencesShowPolitics"
            ) { next in
                await viewModel.setShowPolitics(next)
                onPrefsChanged()
            }
            HStack(spacing: Spacing.s2) {
                Icon(.info, size: 13, color: Theme.Color.appTextMuted)
                Text("These preferences sync across all your devices.")
                    .pantopusTextStyle(.caption)
                    .foregroundStyle(Theme.Color.appTextMuted)
            }
            .padding(.top, Spacing.s5)
            if let toast = viewModel.toastMessage {
                Text(toast)
                    .font(.system(size: 12, weight: .medium))
                    .foregroundStyle(Theme.Color.error)
                    .padding(.top, Spacing.s2)
                    .task(id: toast) {
                        try? await Task.sleep(nanoseconds: 2_500_000_000)
                        viewModel.toastMessage = nil
                    }
            }
        }
        .padding(.horizontal, Spacing.s5)
        .accessibilityIdentifier("pulsePreferencesLoaded")
    }

    private func errorFrame(message: String) -> some View {
        VStack(spacing: Spacing.s3) {
            Icon(.alertCircle, size: 32, color: Theme.Color.error)
            Text("Couldn't load preferences")
                .font(.system(size: 16, weight: .bold))
                .foregroundStyle(Theme.Color.appText)
            Text(message)
                .font(.system(size: 13))
                .foregroundStyle(Theme.Color.appTextSecondary)
                .multilineTextAlignment(.center)
            Button {
                Task { await viewModel.refresh() }
            } label: {
                Text("Try again")
                    .font(.system(size: 14, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
                    .padding(.horizontal, 22)
                    .frame(height: 40)
                    .background(Theme.Color.primary600)
                    .clipShape(Capsule())
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("pulsePreferencesRetry")
        }
        .frame(maxWidth: .infinity)
        .padding(Spacing.s5)
        .accessibilityIdentifier("pulsePreferencesError")
    }

    private func sectionLabel(_ text: String) -> some View {
        Text(text)
            .font(.system(size: 11, weight: .bold))
            .tracking(1)
            .foregroundStyle(Theme.Color.appTextMuted)
            .padding(.top, Spacing.s5)
            .padding(.bottom, Spacing.s2)
    }

    private func preferenceRow(
        title: String,
        body: String,
        isOn: Bool,
        identifier: String,
        onChange: @escaping @MainActor (Bool) async -> Void
    ) -> some View {
        HStack(alignment: .center, spacing: Spacing.s3) {
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.system(size: 15, weight: .semibold))
                    .foregroundStyle(Theme.Color.appText)
                Text(body)
                    .pantopusTextStyle(.caption)
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .fixedSize(horizontal: false, vertical: true)
            }
            Spacer(minLength: Spacing.s2)
            Toggle(
                "",
                isOn: Binding(
                    get: { isOn },
                    set: { next in Task { await onChange(next) } }
                )
            )
            .labelsHidden()
            .tint(Theme.Color.primary600)
            .disabled(viewModel.isSaving)
            .accessibilityLabel(title)
            .accessibilityIdentifier(identifier)
        }
        .padding(.vertical, 14)
        .overlay(alignment: .bottom) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
        }
    }
}

#Preview {
    FeedPreferencesSheet {}
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/feed/FeedPreferencesSheet.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongMethod")

package app.pantopus.android.ui.screens.feed

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Switch
import androidx.compose.material3.SwitchDefaults
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import app.pantopus.android.data.api.models.feed.FeedPreferencesDto
import app.pantopus.android.ui.components.Shimmer
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

/**
 * Pulse Preferences — the bottom sheet behind the Pulse header's
 * preferences control. Mirrors RN
 * `src/components/feed/FeedPreferencesSheet.tsx`: two Place-feed toggles
 * (deals, safety alerts) plus a single political-content toggle that writes
 * both the Place and Connections columns.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FeedPreferencesSheet(
    onClose: () -> Unit,
    onPrefsChanged: () -> Unit = {},
    viewModel: FeedPreferencesViewModel = hiltViewModel(),
) {
    val state by viewModel.state.collectAsStateWithLifecycle()
    val isSaving by viewModel.isSaving.collectAsStateWithLifecycle()
    val toast by viewModel.toastMessage.collectAsStateWithLifecycle()
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    LaunchedEffect(Unit) { viewModel.load() }

    ModalBottomSheet(
        onDismissRequest = onClose,
        sheetState = sheetState,
        containerColor = PantopusColors.appBg,
        modifier = Modifier.testTag("pulsePreferencesSheet"),
    ) {
        Column(modifier = Modifier.fillMaxWidth().padding(horizontal = Spacing.s5)) {
            Row(
                modifier = Modifier.fillMaxWidth().padding(bottom = Spacing.s3),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = "Pulse Preferences",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    color = PantopusColors.appText,
                    modifier = Modifier.semantics { heading() },
                )
                Spacer(modifier = Modifier.weight(1f))
                Box(
                    modifier =
                        Modifier
                            .size(32.dp)
                            .clickable(onClick = onClose)
                            .semantics { contentDescription = "Close preferences" }
                            .testTag("pulsePreferencesClose"),
                    contentAlignment = Alignment.Center,
                ) {
                    PantopusIconImage(
                        icon = PantopusIcon.X,
                        contentDescription = null,
                        size = 18.dp,
                        tint = PantopusColors.appTextSecondary,
                    )
                }
            }
            Box(
                modifier =
                    Modifier
                        .fillMaxWidth()
                        .height(1.dp)
                        .background(PantopusColors.appBorderSubtle),
            )
            when (val s = state) {
                is FeedPreferencesUiState.Loading -> PreferencesLoading()
                is FeedPreferencesUiState.Error ->
                    PreferencesError(message = s.message, onRetry = { viewModel.refresh() })
                is FeedPreferencesUiState.Loaded ->
                    PreferencesLoaded(
                        prefs = s.preferences,
                        isSaving = isSaving,
                        onShowDeals = {
                            viewModel.setShowDeals(it)
                            onPrefsChanged()
                        },
                        onShowAlerts = {
                            viewModel.setShowAlerts(it)
                            onPrefsChanged()
                        },
                        onShowPolitics = {
                            viewModel.setShowPolitics(it)
                            onPrefsChanged()
                        },
                    )
            }
            toast?.let { message ->
                Text(
                    text = message,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Medium,
                    color = PantopusColors.error,
                    modifier = Modifier.padding(top = Spacing.s2).testTag("pulsePreferencesToast"),
                )
            }
            Spacer(modifier = Modifier.height(Spacing.s10))
        }
    }
}

@Composable
private fun PreferencesLoading() {
    Column(
        modifier = Modifier.fillMaxWidth().padding(top = Spacing.s4).testTag("pulsePreferencesLoading"),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        // Shimmer pins its own width, so the skeleton rows use the card
        // width from the A03 frame rather than fillMaxWidth().
        repeat(3) {
            Shimmer(width = 320.dp, height = 56.dp, cornerRadius = Radii.sm)
        }
    }
}

@Composable
private fun PreferencesError(
    message: String,
    onRetry: () -> Unit,
) {
    Column(
        modifier = Modifier.fillMaxWidth().padding(Spacing.s5).testTag("pulsePreferencesError"),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        PantopusIconImage(
            icon = PantopusIcon.AlertCircle,
            contentDescription = null,
            size = 32.dp,
            tint = PantopusColors.error,
        )
        Text(
            text = "Couldn't load preferences",
            fontSize = 16.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.appText,
        )
        Text(text = message, fontSize = 13.sp, color = PantopusColors.appTextSecondary)
        Box(
            modifier =
                Modifier
                    .clip(RoundedCornerShape(Radii.pill))
                    .background(PantopusColors.primary600)
                    .clickable(onClick = onRetry)
                    .padding(horizontal = 22.dp)
                    .height(40.dp)
                    .testTag("pulsePreferencesRetry"),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = "Try again",
                fontSize = 14.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appTextInverse,
            )
        }
    }
}

@Composable
private fun PreferencesLoaded(
    prefs: FeedPreferencesDto,
    isSaving: Boolean,
    onShowDeals: (Boolean) -> Unit,
    onShowAlerts: (Boolean) -> Unit,
    onShowPolitics: (Boolean) -> Unit,
) {
    Column(modifier = Modifier.fillMaxWidth().testTag("pulsePreferencesLoaded")) {
        SectionLabel("PLACE FEED")
        PreferenceRow(
            title = "Show deals",
            body = "Deals and promotions from local businesses",
            checked = !prefs.hideDealsPlace,
            enabled = !isSaving,
            testTag = "pulsePreferencesShowDeals",
            onCheckedChange = onShowDeals,
        )
        PreferenceRow(
            title = "Show safety alerts",
            body = "Crime reports, hazards, and safety warnings",
            checked = !prefs.hideAlertsPlace,
            enabled = !isSaving,
            testTag = "pulsePreferencesShowAlerts",
            onCheckedChange = onShowAlerts,
        )
        SectionLabel("CONTENT")
        PreferenceRow(
            title = "Show political content",
            body =
                "Political posts are hidden by default to keep your feed focused " +
                    "on neighborhood life",
            checked = prefs.showPoliticsPlace,
            enabled = !isSaving,
            testTag = "pulsePreferencesShowPolitics",
            onCheckedChange = onShowPolitics,
        )
        Row(
            modifier = Modifier.fillMaxWidth().padding(top = Spacing.s5),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
        ) {
            PantopusIconImage(
                icon = PantopusIcon.Info,
                contentDescription = null,
                size = 13.dp,
                tint = PantopusColors.appTextMuted,
            )
            Text(
                text = "These preferences sync across all your devices.",
                fontSize = 12.sp,
                color = PantopusColors.appTextMuted,
            )
        }
    }
}

@Composable
private fun SectionLabel(text: String) {
    Text(
        text = text,
        fontSize = 11.sp,
        fontWeight = FontWeight.Bold,
        color = PantopusColors.appTextMuted,
        modifier = Modifier.padding(top = Spacing.s5, bottom = Spacing.s2),
    )
}

@Composable
private fun PreferenceRow(
    title: String,
    body: String,
    checked: Boolean,
    enabled: Boolean,
    testTag: String,
    onCheckedChange: (Boolean) -> Unit,
) {
    Column(modifier = Modifier.fillMaxWidth()) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(vertical = 14.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(Spacing.s3),
        ) {
            Column(modifier = Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text(
                    text = title,
                    fontSize = 15.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.appText,
                )
                Text(text = body, fontSize = 12.sp, color = PantopusColors.appTextSecondary)
            }
            Switch(
                checked = checked,
                onCheckedChange = onCheckedChange,
                enabled = enabled,
                colors =
                    SwitchDefaults.colors(
                        checkedThumbColor = PantopusColors.appTextInverse,
                        checkedTrackColor = PantopusColors.primary600,
                    ),
                modifier =
                    Modifier
                        .semantics { contentDescription = title }
                        .testTag(testTag),
            )
        }
        Box(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .height(1.dp)
                    .background(PantopusColors.appBorderSubtle),
        )
    }
}
